Add support for `mremap(2)` on Linux

Phantomical opened the pull request:
This PR adds adds bindings for mremap(2) on Linux. It introduces the following new methods:
- `Mmap::remap`
- `MmapMut::remap`
- `MmapRaw::remap` and the `RemapOptions` type as a builder for supported options.

The only flag supported in this commit is the `MREMAP_MAYMOVE` flag (as `RemapOptions::may_move`). I have left out `MREMAP_FIXED` and `MREMAP_DONTUNMAP`.

I have marked all the remap functions as unsafe since I think that being able to resize past the end of a file allowing the program to generate SIGBUS signals by accessing the mapped memory probably qualifies as undefined behaviour in rust.

# Open Questions
- This PR ends up duplicating a lot of the tricky pointer arithmetic and memory safety bits elsewhere in the crate. I've held off on doing any refactoring here to try and keep the PR small. If you think it is appropriate I can refactor some of it into helper methods on `MmapInner`.
- Would it be better to remove the `#[cfg(target_os = "linux")]` from `RemapOptions` and only leave it on the `may_move` method?

Closes #81

adamreichold commented:
> The only flag supported in this commit is the MREMAP_MAYMOVE flag (as RemapOptions::may_move).

We do implement `StableDeref` which should be mentioned in the doc comments as its properties are invalidated when the above is used.

> Would it be better to remove the #[cfg(target_os = "linux")] from RemapOptions and only leave it on the may_move method?

Personally, I think it matches up nicely with the only method which uses the type being cfg-ed similarly.

adamreichold commented on src/unix.rs:
```diff
@@ -254,6 +255,63 @@ impl MmapInner {
+        // It is not possible to specify them via RemapOptions but this way if
+        // it is modified in the future then the panic should immediately indicate
+        // that something needs to be changed here.
+        assert_eq!(
```
Since this would be a logic error during development, I think this should just be a `debug_assert!`.

adamreichold commented on src/unix.rs:
```diff
@@ -254,6 +255,63 @@ impl MmapInner {
+        // the pointer and length passed to mmap.
+        //
+        // See the comments in MmapInner::new for relevant details.
+        let alignment = self.ptr() as usize % page_size();
```
> This PR ends up duplicating a lot of the tricky pointer arithmetic and memory safety bits elsewhere in the crate. I've held off on doing any refactoring here to try and keep the PR small. If you think it is appropriate I can refactor some of it into helper methods on MmapInner.

I do think we should factor these computations out and move them closer together so that probability of changing one but not the other is reduced. Not sure how that would look like though...

RazrFalcon replied:
Agree. This code should be moved to a function.

RazrFalcon commented on src/lib.rs:
```diff
@@ -1175,6 +1238,38 @@ impl fmt::Debug for MmapMut {
+#[derive(Copy, Clone, Default, Debug)]
+#[cfg(target_os = "linux")]
+pub struct RemapOptions {
+    pub(crate) flags: libc::c_int,
```
I would suggest using just `may_move: bool` here and resolving the actual value when needed. No need for `pub(crate)` either.

RazrFalcon commented on src/lib.rs:
```diff
@@ -1131,6 +1173,27 @@ impl MmapMut {
+    ///
+    /// See the [`mremap(2)`] man page.
+    ///
+    /// # Safety
```
Add a new line after `# Safety`

RazrFalcon commented on src/unix.rs:
```diff
@@ -254,6 +255,65 @@ impl MmapInner {
+        // It is not possible to specify them via RemapOptions but this way if
+        // it is modified in the future then the panic should immediately indicate
+        // that something needs to be changed here.
+        debug_assert_eq!(
```
This check would not be needed after the `RemapOptions` update.

Phantomical commented:
I've given a shot at refactoring the common pointer handling parts into helper methods. I was hoping to be able to define a `as_raw_parts/from_raw_parts` pair but the special case when the memory map would be length zero made it so that wouldn't work. I think the `adjust_mmap_params/as_mmap_params/from_raw_parts` is an alright alternative.

I have also moved the length checks to `lib.rs`. It seemed to me like `MmapRaw` doesn't need to restrict the length to `isize::MAX` so I have only added checks for `Mmap::remap` and `MmapMut::remap`. I'm not entirely confident about this, though, and would be fine with putting it back if you think it is not appropriate.

RazrFalcon commented:
Oh wow, that's a lot of changes. How about we merge the `mremap` first and then do refactoring is a separate patch? Right now it's hard to figure out what it happening.

RazrFalcon commented on src/lib.rs:
```diff
@@ -1175,6 +1243,62 @@ impl fmt::Debug for MmapMut {
+    }
+
+    pub(crate) fn into_flags(self) -> libc::c_int {
+        match self.may_move {
```
`if` would be fine here.

RazrFalcon commented on src/unix.rs:
```diff
@@ -79,25 +102,78 @@ impl MmapInner {
+        //    we fudge it slightly: a zero-length memory map turns into a
+        //    mapping of length one and can't be told apart outside of this
+        //    method without knowing the original length.
+        match len {
```
And again, `if` would be just fine.

RazrFalcon commented on src/unix.rs:
```diff
@@ -79,25 +102,78 @@ impl MmapInner {
-            }
+    /// Construct this `MmapInner` from its raw components
+    ///
+    /// # Parameters
```
We usually do not have `Parameters`/arguments section in Rust. And it's kinda self-explanatory anyway.

Phantomical commented:
I've reverted the refactoring changes - I agree that it made this a bit too large. I'll open up another PR once this one gets merged with them.

Other than that, I've addressed the remaining comments and I hope things should be straightforward to review now.

adamreichold commented:
> I have also moved the length checks to lib.rs. It seemed to me like MmapRaw doesn't need to restrict the length to isize::MAX so I have only added checks for Mmap::remap and MmapMut::remap. I'm not entirely confident about this, though, and would be fine with putting it back if you think it is not appropriate.

I think the length check should definitely apply to `MmapRaw` as well. The `isize::MAX` limit applies to all allocations in Rust, so I think it would be a huge foot gun, if the raw pointer made available by `MmapRaw` could sometimes not be accessed as a single allocation.

> We do implement StableDeref which should be mentioned in the doc comments as its properties are invalidated when the above is used.

I also still think this trait should be specifically mentioned, for example in the docs of `RemapOptions::may_move()`, as its requirements conflict with moving the mapping. (It is not a soundness issue as `remap` is unsafe, but I think it would still be helpful to make users aware of the conflict.)

Phantomical commented:
> I think the length check should definitely apply to MmapRaw as well.
I have moved the length check back into `MmapInner::remap` so it should apply everywhere now.

> We do implement StableDeref which should be mentioned in the doc comments as its properties are invalidated when the above is used.
I must have missed when addressing the review comments earlier. I have added a note to the docs for `may_move` to this effect. I do think the note is helpful but I do want to mention that I don't think `remap` actually violates the guarantees for `StableDeref`. `stable_deref_trait` implements `StableDeref` for `Vec<T>` which should have the same issue.

RazrFalcon commented:
Looks good now. Thanks! 
Just run `cargo fmt` and we're good to go.
